cpp/src/arrow/compute/kernels/scalar_if_else_test.cc

```diff
@@ -103,7 +118,7 @@ void CheckWithDifferentShapes(const std::shared_ptr<Array>& cond,
   auto len = left->length();
 
   enum { COND_SCALAR = 1, LEFT_SCALAR = 2, RIGHT_SCALAR = 4 };
-  for (int mask = 0; mask < (COND_SCALAR | LEFT_SCALAR | RIGHT_SCALAR); ++mask) {
+  for (int mask = 1; mask <= (COND_SCALAR | LEFT_SCALAR | RIGHT_SCALAR); ++mask) {
     for (int64_t cond_idx = 0; cond_idx < len; ++cond_idx) {
```

Comment:
Speaking of which, if `mask & COND_SCALAR` is false, this will loop `len` times doing the same thing?

Reply:
You could avoid that e.g.:
```c++
std::vector<int64_t> array_indices{-1};  // sentinel
std::vector<int64_t> scalar_indices(len);
std::iota(scalar_indices.begin(), scalar_indices.end(), 0);

auto make_input = [&](const std::shared_ptr<Array>& array, int64_t index,
    Datum* input, Datum* input_bcast, std::string* trace) {
  if (index >= 0) {
     // Use scalar as input, broadcast it for expected result computation
     ASSERT_OK_AND_ASSIGN(*input, array->GetScalar(index));
     ASSERT_OK_AND_ASSIGN(*input_bcast, MakeArrayFromScalar(*input->scalar(), len));
     *trace += "...":
  } else {
     // Use array as input
     *input = *input_bcast = array;
     *trace += "=Array";
  }
};

for (int mask = 1; mask <= (COND_SCALAR | LEFT_SCALAR | RIGHT_SCALAR); ++mask) {
  for (const auto cond_idx : (mask & COND_SCALAR) ? scalar_indices : array_indices) {
    Datum cond_in, cond_bcast;
    std::string trace_cond = "Cond";
    make_input(cond, cond_idx, &cond_in, &cond_bcast, &trace_cond);
    for (const auto left_idx : (mask & LEFT_SCALAR) ? scalar_indices : array_indices) {
      // etc.
```

Reply:
Thanks, this makes things much more readable here.